lib/decryptor.rb

```diff
@@ -78,34 +78,65 @@ def shift_amount(encrypted_char, end_char)
   end
 
   def find_key(encrypted_message, date)
-    key  = ""
-    offsets = date_to_offsets(date)
-    find_shifts(encrypted_message, date).each_with_index do |shift, index|
+    combine_keys(find_working_keys(encrypted_message, date))
+  end
+
+  def combine_keys(keys)
+    keys.each_with_index.map do |key, index|
       if index == 0
-        key.concat(find_first_key(shift, offsets[index]))
+        key
       else
-        find_next_key_value(key, index, offsets[index], shift)
+        key[1]
       end
-    end
-    key
+    end.join
   end
 
-  def find_first_key(shift, offset)
-    if ((shift - offset) % 27).to_s.length == 1
-      "0" + ((shift - offset) % 27).to_s
-    else
-      ((shift - offset) % 27).to_s
+  def find_working_keys(encrypted_message, date)
+    potential_keys = find_all_potential_keys(encrypted_message, date)
+    potential_keys.first.each do |key_1|
+      next if !key_is_valid?(potential_keys[1], key_1)
+      key_2 = find_next_valid_key(potential_keys[1], key_1)
+      next if !key_is_valid?(potential_keys[2], key_2)
+      key_3 = find_next_valid_key(potential_keys[2], key_2)
+      next if !key_is_valid?(potential_keys[3], key_3)
+      key_4 = find_next_valid_key(potential_keys[3], key_3)
+      return [key_1, key_2, key_3, key_4]
```

Comment:
This method iterated over the first set of potential keys until it finds one that has a valid pair in the next set of potential keys. It then checks if the valid pair also has a valid pair in the next set and it will do this until it finds one with a valid chain of 4 keys.